projects/stretch_continual/stretch_continual/envs/stretch_demo_base_env.py

```diff
@@ -0,0 +1,312 @@
+                                        self._trajectory_cache["linked_files"][
+                                            f"file_{h5_id}"
+                                        ] = self._recursively_load_h5py(
+                                            trajectory_data=trajectory_file
+                                        )
+
+                                h5_id += 1
+
+                    break
+                except (BlockingIOError, OSError) as e:
+                    if try_id == max_tries - 1:
+                        print(
+                            f"Failed to open or read {output_path} with error {e} after {max_tries} attempts."
+                        )
+                        raise e
+
+                    print(
+                        f"Failing to open or read {output_path} with error {e}. Retrying..."
+                    )
+
+        if cache:
+            demo_data = CacheContainer(self._trajectory_cache)
+        else:
+            demo_data = h5py.File(output_path, "r")
+
+        return demo_data
+
+    def randomly_select_traj_from_dir(self, directory, only_key_frames, cache):
+        max_tries = 25
+
+        for try_id in range(max_tries):
+            try:
+                with tempfile.NamedTemporaryFile(
+                    dir=directory
+                ) as temp_aggregation_file:
+                    with self.load_all_h5_from_dir(
+                        directory,
+                        only_key_frames,
+                        temp_aggregation_file,
+                        cache=cache,
+                    ) as h5_file:
+                        aggregated_h5 = h5_file["linked_files"]
+                        traj_counts = [
+                            (file_key, len(item.keys()))
+                            if item is not None
+                            else (None, 0)
+                            for file_key, item in aggregated_h5.items()
+                        ]
+                        total_trajs = np.array(
+                            [count[1] for count in traj_counts]
+                        ).sum()
+                        traj_id = np.random.randint(total_trajs)
+                        curr_id = 0
+                        traj = None
+
+                        for file_id, count in traj_counts:
+                            # If this file contains our selected trajectory (i.e. it's within the count for this file), grab it
+                            if curr_id + count > traj_id:
+                                traj = aggregated_h5[file_id][f"{traj_id - curr_id}"]
+                                break
+
+                            curr_id += count
+
+                break
+            except (BlockingIOError, OSError, KeyError) as e:
+                time.sleep(0.1)
+
+                if try_id == max_tries - 1:
+                    print(f"Ran out of re-attempts to load the h5.")
+                    raise e
+
+        return traj
+
+    def gripper_fk(self, model, pose):
```

Comment:
type hints + why do we need a separate fk function? should it not be rolled into stretch kinematics class?

Reply:
This is because I want to be sure all the continual envs to be able to use consistent fk/ik, and I was finding that I was switching between fk/ik implementations/APIs. So all these envs use the same method, and then I only have to switch what I'm doing in this one place. (E.g. when I found that copying the pose was helpful, doing it in just one place was nice.) Assuming the API is stable I can probably remove it...but it also means that as I generalize this code (e.g. to sim), I should only need to override this, if it's different.

I'm going through now and figuring out the best way to generalize stuff, so...I might commit this, and then do another PR with changes to better enable generalization?

Reply:
This makes sense! We can keep it especially for this PR
is there a way to use a similar wrapper for the sim experiments in the future?